### PR TITLE
[FAD-95] IPFS

### PR DESCRIPTION
**Pull request by sachasmart-weavik**
### Description

> Containerized IPFS and file uploader. 
Read the packages/ipfs/README.md and I can demo it on Monday before you two look at the PR. 

### Checklist

- [ ] `CHANGELOG` has been updated (_if appropriate_)
- [ ] Environment variables updated (_if appropriate_)

### CODE REVIEW

File: packages/client3/components/design-system/file-upload/file-upload.tsx [hunk 1]
```diff
@@ -1,50 +1,120 @@
...
+        ipfs.add(binaryStr).then((file) => {
+          console.log(file); //IPFS RETURN
+        });
+      }
+    };
+    reader.readAsArrayBuffer(fileToUpload);
+  } catch (error) {
+    console.log(error);
+  }
+};
```
Comment (v1n33th) [inline packages/client3/components/design-system/file-upload/file-upload.tsx:29]: Would suggest converting the file reader method to [promise](https://simon-schraeder.de/posts/filereader-async/) and then extracting the addFile to a utils/services class to aid reusability. Also would suggest using camelCase for function names

File: packages/client3/components/design-system/file-upload/file-upload.tsx [hunk 2]
```diff
@@ -1,50 +1,120 @@
...
+
     let _errorMessage = errorMessage;
     if (formContext !== null) {
       _errorMessage = formContext.formState.errors[name]?.message;
     }
     const isError = !!_errorMessage;
 
+    useCallback((acceptedFile) => {
+      AddFile(acceptedFile[0]);
+    }, []);
```
Comment (v1n33th) [inline packages/client3/components/design-system/file-upload/file-upload.tsx:44]: Is this useCallback function called as it's not assigned to any variables to use?

File: packages/client3/ipfs/README.md [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+- Documentation (horrible): https://ipfs.io/docs
+
+# SETUP
+
+1. Run `make ipfs` to run spin up IPFS container and daemon. Access IPFS WebUI at http://localhost:5001/webui
+2. First test:
+   1. Upload a file with the File Uploader
+   2. Look in the console and you will see a: `Qm...` hash
+   3. go to http://localhost:5001/webui and enter the hash into the search bar
+   4. Magic.
```
Comment (v1n33th) [inline packages/client3/ipfs/README.md:13]: lol

File: packages/client3/components/design-system/file-upload/file-upload.tsx [hunk 4]
```diff
@@ -1,50 +1,120 @@
...
+    };
+    reader.readAsArrayBuffer(fileToUpload);
+  } catch (error) {
+    console.log(error);
+  }
+};
 
 export const FileUpload = forwardRef<HTMLInputElement, FileUploadProps>(
   function FileUpload({ name, errorMessage, disabled }, ref) {
+    const [files, setFiles] = useState([] as any);
```
Comment (v1n33th) [inline packages/client3/components/design-system/file-upload/file-upload.tsx:33]: Does the component support multiple files?

**Review by v1n33th — CHANGES_REQUESTED** [review]
Awesome...Some minor suggestions

File: packages/client3/components/design-system/file-upload/file-upload.tsx [hunk 5]
```diff
@@ -1,50 +1,118 @@
...
+    const coverImage = file[0]?.preview;
+
+    useEffect(() => {
+      if (file.length > 0) {
+        if (isError) {
+          return;
+        }
+        addFile(file[0]);
+      }
+    });
```
Comment (v1n33th) [inline packages/client3/components/design-system/file-upload/file-upload.tsx:49]: This useEffect runs on every render as it has no dependency. Is this intended?

File: packages/client3/components/design-system/file-upload/file-upload.tsx [hunk 6]
```diff
@@ -1,50 +1,118 @@
...
+
+    useEffect(
+      () => () => {
```
Comment (v1n33th) [inline packages/client3/components/design-system/file-upload/file-upload.tsx:52]: Is this by mistake or required?

File: packages/client3/ipfs/utils/services/file-upload.ts [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+      return await ipfs.add(binaryStr).then((file) => {
+        pinFilesToIPFS(file.path);
+        console.log(
+          "content is availabe at:",
+          ipfsBaseUrl + file.cid
+          // future Sacha problem, but add external pinning - tested with cloudflare-ipfs (50GB/month free) and works
+        );
+      });
+    }
+    return await fileReader.readAsArrayBuffer(fileToUpload);
```
Comment (v1n33th) [inline packages/client3/ipfs/utils/services/file-upload.ts:28]: Not sure. But i don't think `readAsArrayBuffer` is a async function and can be waited. Maybe promisify `readAsArrayBuffer` method and use await

**Review by v1n33th — APPROVED** [review]
Just some minor suggestions

**Comment by sachasmart-weavik** [comment]
Closing due to the changes when implementing [[FAD-108] ](https://github.com/weavik/free-artist-dao/pull/33)

**Pull request closed without merging**